File: tests/01-numpy_arrays/004-shape_attr/test_shape_attr.nim
import strutils  # `%`
import pymod
import pymodpkg/miscutils
import pymodpkg/pyarrayobject


# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnDimensionsAsTuple1D*(arr: ptr PyArrayObject): tuple[a: npy_intp] {.exportpy.} =
proc returnDimensionsAsTuple1D*(arr: ptr PyArrayObject): tuple[a: int] {.exportpy.} =
  let expected_nd = 1
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (a: arr.dimensions[0])

# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnShapeAsTuple1D*(arr: ptr PyArrayObject): tuple[a: npy_intp] {.exportpy.} =
proc returnShapeAsTuple1D*(arr: ptr PyArrayObject): tuple[a: int] {.exportpy.} =
  let expected_nd = 1
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (a: arr.shape[0])


# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnDimensionsAsTuple2D*(arr: ptr PyArrayObject): tuple[a, b: npy_intp] {.exportpy.} =
proc returnDimensionsAsTuple2D*(arr: ptr PyArrayObject): tuple[a, b: int] {.exportpy.} =
  let expected_nd = 2
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.dimensions[0], arr.dimensions[1])

# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnShapeAsTuple2D*(arr: ptr PyArrayObject): tuple[a, b: npy_intp] {.exportpy.} =
proc returnShapeAsTuple2D*(arr: ptr PyArrayObject): tuple[a, b: int] {.exportpy.} =
  let expected_nd = 2
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.shape[0], arr.shape[1])


# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnDimensionsAsTuple3D*(arr: ptr PyArrayObject): tuple[a, b, c: npy_intp] {.exportpy.} =
proc returnDimensionsAsTuple3D*(arr: ptr PyArrayObject): tuple[a, b, c: int] {.exportpy.} =
  let expected_nd = 3
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.dimensions[0], arr.dimensions[1], arr.dimensions[2])

# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnShapeAsTuple3D*(arr: ptr PyArrayObject): tuple[a, b, c: npy_intp] {.exportpy.} =
proc returnShapeAsTuple3D*(arr: ptr PyArrayObject): tuple[a, b, c: int] {.exportpy.} =
  let expected_nd = 3
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.shape[0], arr.shape[1], arr.shape[2])


# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnDimensionsAsTuple4D*(arr: ptr PyArrayObject): tuple[a, b, c, d: npy_intp] {.exportpy.} =
proc returnDimensionsAsTuple4D*(arr: ptr PyArrayObject): tuple[a, b, c, d: int] {.exportpy.} =
  let expected_nd = 4
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.dimensions[0], arr.dimensions[1], arr.dimensions[2], arr.dimensions[3])

# FIXME:  Technically, this really should return `npy_intp` rather than `int`.
# Yes, the two types are identical (a signed integer that is the same size as
# a pointer), but still... technically...
#proc returnShapeAsTuple4D*(arr: ptr PyArrayObject): tuple[a, b, c, d: npy_intp] {.exportpy.} =
proc returnShapeAsTuple4D*(arr: ptr PyArrayObject): tuple[a, b, c, d: int] {.exportpy.} =
  let expected_nd = 4
  if arr.nd != expected_nd:
    let msg = "expected input array of ndim=$1, received ndim=$2" % [$expected_nd, $arr.nd]
    raise newException(ValueError, msg)
  else:
    result = (arr.shape[0], arr.shape[1], arr.shape[2], arr.shape[3])


initPyModule("",
    returnDimensionsAsTuple1D, returnShapeAsTuple1D,
    returnDimensionsAsTuple2D, returnShapeAsTuple2D,
    returnDimensionsAsTuple3D, returnShapeAsTuple3D,
    returnDimensionsAsTuple4D, returnShapeAsTuple4D)

File: tests/01-numpy_arrays/004-shape_attr/test_shape_attr.py
import array_utils
import numpy
import pytest


def test_0_compile_pymod_test_mod(pmgen_py_compile):
        pmgen_py_compile(__name__)


# TODO
#@pytest.mark.parametrize("array_shape", array_utils.all_0d_array_shapes)
#@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
#def test_returnDimensionsAsTuple0D(pymod_test_mod, seeded_random_number_generator,
#        array_shape, input_type):
#    arg = array_utils.get_random_Nd_array_of_shape_and_type(array_shape, input_type)
#    expectedDimensions = arg.shape
#    expectedShape = arg.shape
#
#    # FIXME: To be implemented.
#    resDimensions = pymod_test_mod.returnDimensionsAsTuple0D(arg)
#    assert resDimensions == expectedDimensions
#
#    # FIXME: To be implemented.
#    resShape = pymod_test_mod.returnShapeAsTuple0D(arg)
#    assert resShape == expectedShape


@pytest.mark.parametrize("array_shape",
        array_utils.small_1d_array_shapes + array_utils.zero_size_1d_array_shapes)
@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple1D1(pymod_test_mod, seeded_random_number_generator,
        array_shape, input_type):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple1D(arg)
    # FIXME:  Currently, Pymod incorrectly unwraps single-element-tuple return-types.
    # Thus, `resDimensions` should be a tuple-of-single-int, but instead it's an int.
    # We will fix this soon...
    resDimensions = (resDimensions,)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple1D(arg)
    # FIXME:  Currently, Pymod incorrectly unwraps single-element-tuple return-types.
    # Thus, `resShape` should be a tuple-of-single-int, but instead it's an int.
    # We will fix this soon...
    resShape = (resShape,)
    assert resShape == expectedShape


@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple1D2(pymod_test_mod, input_type, random_1d_array_size):
    arg = array_utils.get_random_1d_array_of_size_and_type(random_1d_array_size, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple1D(arg)
    # FIXME:  Currently, Pymod incorrectly unwraps single-element-tuple return-types.
    # Thus, `resDimensions` should be a tuple-of-single-int, but instead it's an int.
    # We will fix this soon...
    resDimensions = (resDimensions,)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple1D(arg)
    # FIXME:  Currently, Pymod incorrectly unwraps single-element-tuple return-types.
    # Thus, `resShape` should be a tuple-of-single-int, but instead it's an int.
    # We will fix this soon...
    resShape = (resShape,)
    assert resShape == expectedShape


@pytest.mark.parametrize("array_shape",
        array_utils.small_2d_array_shapes + array_utils.zero_size_2d_array_shapes)
@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple2D1(pymod_test_mod, seeded_random_number_generator,
        array_shape, input_type):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple2D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple2D(arg)
    assert resShape == expectedShape


@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple2D2(pymod_test_mod, input_type, random_2d_array_shape):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(random_2d_array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple2D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple2D(arg)
    assert resShape == expectedShape


@pytest.mark.parametrize("array_shape",
        array_utils.small_3d_array_shapes + array_utils.zero_size_3d_array_shapes)
@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple3D1(pymod_test_mod, seeded_random_number_generator,
        array_shape, input_type):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple3D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple3D(arg)
    assert resShape == expectedShape


@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple3D2(pymod_test_mod, input_type, random_3d_array_shape):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(random_3d_array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple3D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple3D(arg)
    assert resShape == expectedShape


@pytest.mark.parametrize("array_shape", array_utils.small_4d_array_shapes)
@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple4D1(pymod_test_mod, seeded_random_number_generator,
        array_shape, input_type):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple4D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple4D(arg)
    assert resShape == expectedShape


@pytest.mark.parametrize("input_type", array_utils.all_supported_numpy_types)
def test_returnDimensionsAsTuple4D2(pymod_test_mod, input_type, random_4d_array_shape):
    arg = array_utils.get_random_Nd_array_of_shape_and_type(random_4d_array_shape, input_type)
    expectedDimensions = arg.shape
    expectedShape = arg.shape

    resDimensions = pymod_test_mod.returnDimensionsAsTuple4D(arg)
    assert resDimensions == expectedDimensions

    resShape = pymod_test_mod.returnShapeAsTuple4D(arg)
    assert resShape == expectedShape

